### ISSUE TITLE
[Filestream] Filestream add flag to only terminate the input after reaching EOF on all files

### ISSUE THREAD

**Issue by belimawr** [elastic/beats#40447]
When Filebeat is running with autodiscover enabled it can happen that a pod is terminated, autodiscover signals to terminate the input, but the files have not been read until EOF. Terminating the input on this state will lead to the file not being fully ingested.

There are at least a couple of common cases where this happens:
 - Short lived pods
 - High load environments where Filebeat might experience back pressure from the output

Currently the best way to avoid data loss on those situations is to set [`close.on_state_change.removed: false`](https://www.elastic.co/guide/en/beats/filebeat/current/filebeat-input-filestream.html#filebeat-input-filestream-close-removed) in the input configuration and [`cleanup_timeout`](https://www.elastic.co/guide/en/beats/filebeat/current/configuration-autodiscover.html#_kubernetes) for the autodiscover configuration to a value that will give Filebeat enough time to ingest all data.

**Comment by AndersonQ** [comment]
@nimarezainia, @cmacknz, @belimawr 

I'm trying to understand he correct precedence of this "only terminate filestream after EOF".

 - we're actually referring to reading a file. That means, we only stop reading the file (close the harvester) after EOF
 - all `close.on_state_change.*` need to wait until EOF is reached
 - [`close.reader.after_interval`](https://www.elastic.co/docs/reference/beats/filebeat/filebeat-input-filestream#filebeat-input-filestream-close-timeout) takes precedence over "wait until EOF is reached"

is it?

**Comment by AndersonQ** [comment]
or is [`close.reader.after_interval`](https://www.elastic.co/docs/reference/beats/filebeat/filebeat-input-filestream#filebeat-input-filestream-close-timeout) incompatible with "wait until EOF"

**Comment by belimawr** [comment]
Because the goal here is to ingest files until the end even when the input is supposed to shut down (that's driven by the Kubernetes autodiscover behaviour), I'd say:
 - "only terminate filestream after EOF" is incompatible with settings like:
     - `close.reader.after_interval`
     - `close.on_state_change.renamed: true`
     - `close.on_state_change.removed: true`

Instead of thinking about precedence, I'd not allow conflicting settings to be enabled at the same time.


One thing to keep in mind is that if Filebeat is shutting down, then all Filestream inputs will be signalled to shutdown, so we probably need a timeout for this wait to prevent Filebeat hanging for ever.

Think of this as a "safety feature": if the input is signalled to shutdown, it must ensure all files are read until EOF before shutting down. In that case we can assume that no data will be added to the file after the shutdown signal is sent, so reaching EOF at any point means the file has been fully ingested and no more data will ever be added to it.

Answering your question:
> we're actually referring to reading a file. That means, we only stop reading the file (close the harvester) after EOF

Yes, we're referring to all files in the input. This applies to both: the harvester and the input itself.

**Comment by AndersonQ** [comment]
> * "only terminate filestream after EOF" is incompatible with settings like:
>   
>   * `close.reader.after_interval`
>   * `close.on_state_change.renamed: true`
>   * `close.on_state_change.removed: true`
> 
> Instead of thinking about precedence, I'd not allow conflicting settings to be enabled at the same time.

yeah, that makes sense and makes thing simpler. I'll go the incompatibility route.

btw @belimawr, do you have a way to reproduce the problematic behaviour?

The obvious test it to have a large enough file, start filestream, stop it and ensure the file is read to completion and another where it timeout.
But a more realistic scenario is interesting. So far, using kind rather naively , the logs stay there after the pods are deleted. If you know how to reproduce this behaviour on k8s, let me know

**Comment by belimawr** [comment]
> btw @belimawr, do you have a way to reproduce the problematic behaviour?

Not really because it's caused by correct timing and interactions between the autodiscover and Filestream.

> The obvious test it to have a large enough file, start filestream, stop it and ensure the file is read to completion and another where it timeout.

That sounds good, another trick is to block the output, so the file is not read until EOF and it gives you some time/control, specially if using an integration test.

> But a more realistic scenario is interesting. So far, using kind rather naively , the logs stay there after the pods are deleted. If you know how to reproduce this behaviour on k8s, let me know

The problem here is not the file being deleted, is the pod stopping and that making the Filestream input stop before reading the file until EOF. The file being removed is easily circumvented by setting `close.on_state_change.removed: false`.

**Comment by AndersonQ** [comment]
> > But a more realistic scenario is interesting. So far, using kind rather naively , the logs stay there after the pods are deleted. If you know how to reproduce this behaviour on k8s, let me know
> 
> The problem here is not the file being deleted, is the pod stopping and that making the Filestream input stop before reading the file until EOF. The file being removed is easily circumvented by setting `close.on_state_change.removed: false`.

oh yeah, my bad here

**Comment by cmacknz** [comment]
Continuing on a theme here - how does filelog handle this problem when run in the collector? More specific to us - how does filestream handle this when run in the collector? Does anything change to help here?

**Comment by AndersonQ** [comment]
> More specific to us - how does filestream handle this when run in the collector?

@cmacknz why do you think it would be different? Unless I'm mistaken or I missed something in your question, filebeat will behave the same, stand alone or in otel mode.

**Comment by AndersonQ** [comment]
looking at [filelog docs](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/stanza/fileconsumer/design.md#shutdown-logic), it stops in the middle of the file:

> ### Shutdown Logic
>
>When the operator shuts down, the following occurs:
>- If a poll cycle is not currently underway, the operator simply closes any open files.
>- Otherwise, the current poll cycle is signaled to stop immediately, which in turn signals all >Readers to
>  stop immediately.
>    - If a Reader is idle or in between log entries, it will return immediately. Otherwise it >will return
>      after consuming one final log entry.
>    - Once all Readers have stopped, the remainder of the poll cycle completes as usual, which >includes
>      the steps labeled `Closing`, `Archiving`, `Pruning`, and `Persistence`.
>
>The net effect of the shut down routine is that all files are checkpointed in a normal manner
(i.e. not in the middle of a log entry), and all checkpoints are persisted.

also, in the code, if the context is cancelled, it does not attempt to keep reading: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/b1d8afc3e2c2d7923eb2511d4b088571c65fdb88/pkg/stanza/fileconsumer/internal/reader/reader.go#L228-L251

**Comment by AndersonQ** [comment]
Also, I didn't see configuration to wait before triggering the shutdown or anything in the code, I looked at, having a similar behaviour
